File: src/common/ipc-types.ts
import { PassiveIncomeType } from 'dobkap/lib/passive-income'
import type { BrowserWindow } from 'electron'
import { HolidayConf } from './holiday-conf'
import { DateString } from './primitive-types'

export type ReportStatus = 'init' | 'processed'

export type Report = {
  id: number,
  importerId: number | null,
  mailboxId: number,
  mailboxMessageId: number,
  reportName: string,
  status: ReportStatus,
}

export type FilingStatus = 'init' | 'filed' | 'paid'

export type Filing = {
  id: number,
  reportId: number,
  type: PassiveIncomeType,
  status: FilingStatus,
  payingEntity: string,
  filingDeadline: DateString,
  taxPayable: number
}

export type TaxpayerProfile = {
  id: number,
  jmbg: string,
  fullName: string,
  streetAddress: string,
  opstinaCode: string,
  phoneNumber: string,
  emailAddress: string,
}

export type Mailbox = {
  id: number,
  emailAddress: string,
  emailPassword: string,
  imapHost: string,
  imapPort: number,
  cursor: MailboxCursor,
}

export type ReportJobMessage = {
  type: 'report'
  from: string
  subject: string
  attachmentName: string
}

export type ErrorJobMessage = {
  type: 'error',
  message: string
}

export type SuccessJobMessage = {
  type: 'success',
  message: string
}

export type JobMessage = ReportJobMessage | ErrorJobMessage | SuccessJobMessage

export type Job = {
  id: number
  completed: boolean
  canceled: boolean
  messages: JobMessage[]
}

export type CreateJobResult = {
  id: number
}

export type MailboxCursor = {
  type: 'date'
  dateString: DateString
} | {
  type: 'uid'
  lastSeenUid: number
  lastSeenDatetime: number
}

export type FilingFilter = 'all' | 'unpaid'

export type TechnicalConf = {
  mexicoBdmToken: string
  holidayConf: HolidayConf
}

export type ImporterType = 'IbkrCsv'

export type Importer = {
  id: number
  name: string
  type: ImporterType
  taxpayerProfileId: number
  mailboxId: number
  fromFilter: string
  subjectFilter: string
  paymentNotes: string
  attachmentRegex: string
}

type IpcTypeDef<N extends string, A, B> = {
  Name: N
  Arg: A
  Result: B
  ClientFn: (arg: A) => Promise<B>
  HandlerFn: (req: {
    event: Electron.IpcMainInvokeEvent,
    browserWindow: BrowserWindow,
    arg: A,
  }) => Promise<B>
}

export type OpenUrl = IpcTypeDef<'open-url', string, void>

export type RunSql = IpcTypeDef<
  'run-sql',
  string,
  {
    columns: Array<{ name: string }>
    rows: Array<Array<any>>
  }
>

export type GetReports = IpcTypeDef<
  'get-reports',
  void,
  Array<Report>
>
export type DeleteReport = IpcTypeDef<
  'delete-report',
  number,
  void
>
export type GetFilings = IpcTypeDef<
  'get-filings',
  void,
  Array<Filing>
>
export type UpdateFiling = IpcTypeDef<
  'update-filing',
  Filing,
  void
>
export type ExportFiling = IpcTypeDef<
  'export-filing',
  number,
  void
>
export type DeleteFiling = IpcTypeDef<
  'delete-filing',
  number,
  void
>
export type GetTaxpayerProfile = IpcTypeDef<
  'get-taxpayer-profile',
  void,
  TaxpayerProfile
>
export type UpdateTaxpayerProfile = IpcTypeDef<
  'update-taxpayer-profile',
  TaxpayerProfile,
  void
>
export type GetMailbox = IpcTypeDef<
  'get-mailbox',
  void,
  Mailbox
>
export type UpdateMailbox = IpcTypeDef<
  'update-mailbox',
  Mailbox,
  void
>
export type CreateJob = IpcTypeDef<
  'create-job',
  void,
  CreateJobResult
>
export type GetJob = IpcTypeDef<
  'get-job',
  number,
  Job | null
>
export type CancelJob = IpcTypeDef<
  'cancel-job',
  number,
  void
>
export type GetTechnicalConf = IpcTypeDef<
  'get-technical-conf',
  void,
  TechnicalConf
>
export type UpdateTechnicalConf = IpcTypeDef<
  'update-technical-conf',
  TechnicalConf,
  void
>
export type ImportHolidayConf = IpcTypeDef<
  'import-holiday-conf',
  void,
  HolidayConf | null
>
export type GetImporters = IpcTypeDef<
  'get-importers',
  void,
  Array<Importer>
>
export type UpdateImporter = IpcTypeDef<
  'update-importer',
  Importer,
  void
>
export type DeleteImporter = IpcTypeDef<
  'delete-importer',
  number,
  void
>
export type CreateImporter = IpcTypeDef<
  'create-importer',
  Importer,
  void
>

export type IpcMethods = {
  openUrl: OpenUrl
  runSql: RunSql
  getReports: GetReports
  deleteReport: DeleteReport
  getFilings: GetFilings
  updateFiling: UpdateFiling
  exportFiling: ExportFiling
  deleteFiling: DeleteFiling
  getTaxpayerProfile: GetTaxpayerProfile
  updateTaxpayerProfile: UpdateTaxpayerProfile
  getMailbox: GetMailbox
  updateMailbox: UpdateMailbox
  createJob: CreateJob
  getJob: GetJob
  cancelJob: CancelJob
  getTechnicalConf: GetTechnicalConf
  updateTechnicalConf: UpdateTechnicalConf
  importHolidayConf: ImportHolidayConf
  getImporters: GetImporters
  updateImporter: UpdateImporter
  deleteImporter: DeleteImporter
  createImporter: CreateImporter
}

export type IpcHandlerFns = { [k in keyof IpcMethods]: IpcMethods[k]['HandlerFn']}
export type IpcContextApi = { [k in keyof IpcMethods]: IpcMethods[k]['ClientFn']}
export type IpcNames = { [k in keyof IpcMethods]: IpcMethods[k]['Name']}

const ipcNames: IpcNames = {
  openUrl: 'open-url',
  runSql: 'run-sql',
  getReports: 'get-reports',
  deleteReport: 'delete-report',
  getFilings: 'get-filings',
  updateFiling: 'update-filing',
  exportFiling: 'export-filing',
  deleteFiling: 'delete-filing',
  getTaxpayerProfile: 'get-taxpayer-profile',
  updateTaxpayerProfile: 'update-taxpayer-profile',
  getMailbox: 'get-mailbox',
  updateMailbox: 'update-mailbox',
  createJob: 'create-job',
  getJob: 'get-job',
  cancelJob: 'cancel-job',
  getTechnicalConf: 'get-technical-conf',
  updateTechnicalConf: 'update-technical-conf',
  importHolidayConf: 'import-holiday-conf',
  getImporters: 'get-importers',
  updateImporter: 'update-importer',
  deleteImporter: 'delete-importer',
  createImporter: 'create-importer',
}

export const foreachIpc = (cb: (ipcKey: keyof IpcNames, ipcName: string) => void): void => {
  Object.keys(ipcNames).forEach((ipcKey: keyof IpcNames) => {
    const ipcName  = ipcNames[ipcKey]
    cb(ipcKey, ipcName)
  })
}

File: src-tauri/src/ipc_types.rs
use crate::error::{DkaError, DkaResult};
use serde::{self, Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HolidayConf {
    #[serde(rename = "holidayRangeStart")]
    pub holiday_range_start: String,
    #[serde(rename = "holidayRangeEnd")]
    pub holiday_range_end: String,
    pub holidays: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TechnicalConf {
    #[serde(rename = "holidayConf")]
    pub holiday_conf: HolidayConf,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum MailboxCursor {
    #[serde(rename = "date")]
    Date {
        #[serde(rename = "dateString")]
        date_string: String,
    },
    #[serde(rename = "uid")]
    Uid {
        #[serde(rename = "lastSeenUid")]
        last_seen_uid: i64,
        #[serde(rename = "lastSeenDatetime")]
        last_seen_datetime: i64,
    },
}

impl MailboxCursor {
    pub fn to_db_string(&self) -> String {
        match self {
            MailboxCursor::Date { date_string } => format!("date,{}", date_string),
            MailboxCursor::Uid {
                last_seen_uid,
                last_seen_datetime,
            } => {
                format!("uid,{},{last_seen_datetime}", last_seen_uid)
            }
        }
    }
    pub fn from_db_string(s: &str) -> DkaResult<Self> {
        let parts: Vec<&str> = s.split(',').collect();
        match parts.as_slice() {
            ["date", date_str] => Ok(MailboxCursor::Date {
                date_string: (*date_str).to_string(),
            }),
            ["uid", uid_str, dt_str] => {
                let uid = uid_str
                    .parse::<i64>()
                    .map_err(|_e| DkaError::Generic("could not parse uid".into()))?;
                let dt = dt_str
                    .parse::<i64>()
                    .map_err(|_e| DkaError::Generic("could not parse dt".into()))?;
                Ok(MailboxCursor::Uid {
                    last_seen_uid: uid,
                    last_seen_datetime: dt,
                })
            }
            _ => Err(DkaError::Generic("invalid mailbox cursor format".into())),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Mailbox {
    #[serde(rename = "id")]
    pub id: i32,
    #[serde(rename = "emailAddress")]
    pub email_address: String,
    #[serde(rename = "emailPassword")]
    pub email_password: String,
    #[serde(rename = "imapHost")]
    pub imap_host: String,
    #[serde(rename = "imapPort")]
    pub imap_port: i32,
    #[serde(rename = "cursor")]
    pub cursor: MailboxCursor,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaxpayerProfile {
    #[serde(rename = "id")]
    pub id: i32,
    #[serde(rename = "jmbg")]
    pub jmbg: String,
    #[serde(rename = "fullName")]
    pub full_name: String,
    #[serde(rename = "streetAddress")]
    pub street_address: String,
    #[serde(rename = "opstinaCode")]
    pub opstina_code: String,
    #[serde(rename = "phoneNumber")]
    pub phone_number: String,
    #[serde(rename = "emailAddress")]
    pub email_address: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Importer {
    #[serde(rename = "id")]
    pub id: i32,
    #[serde(rename = "name")]
    pub name: String,
    #[serde(rename = "reportType")]
    pub report_type: String,
    #[serde(rename = "taxpayerProfileId")]
    pub taxpayer_profile_id: i32,
    #[serde(rename = "mailboxId")]
    pub mailbox_id: i32,
    #[serde(rename = "fromFilter")]
    pub from_filter: String,
    #[serde(rename = "subjectFilter")]
    pub subject_filter: String,
    #[serde(rename = "paymentNotes")]
    pub payment_notes: String,
    #[serde(rename = "attachmentRegex")]
    pub attachment_regex: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Report {
    #[serde(rename = "id")]
    pub id: i32,
    #[serde(rename = "type")]
    pub _type: String,
    #[serde(rename = "importerId")]
    pub importer_id: Option<i32>,
    #[serde(rename = "mailboxId")]
    pub mailbox_id: i32,
    #[serde(rename = "mailboxMessageId")]
    pub mailbox_message_id: i32,
    #[serde(rename = "reportName")]
    pub report_name: String,
    #[serde(rename = "status")]
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Filing {
    #[serde(rename = "id")]
    pub id: i32,
    #[serde(rename = "reportId")]
    pub report_id: i32,
    #[serde(rename = "type")]
    pub _type: String,
    #[serde(rename = "status")]
    pub status: String,
    #[serde(rename = "payingEntity")]
    pub paying_entity: String,
    #[serde(rename = "filingDeadline")]
    pub filing_deadline: String,
    #[serde(rename = "taxPayable")]
    pub tax_payable: i64,
    #[serde(rename = "taxPaymentReference")]
    pub tax_payment_reference: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum JobMessage {
    #[serde(rename = "report")]
    Report {
        #[serde(rename = "from")]
        from: String,
        #[serde(rename = "subject")]
        subject: String,
        #[serde(rename = "attachmentName")]
        attachment_name: String,
    },
    #[serde(rename = "error")]
    Error {
        #[serde(rename = "message")]
        message: String,
    },
    #[serde(rename = "success")]
    Success {
        #[serde(rename = "message")]
        message: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Job {
    #[serde(rename = "id")]
    pub id: i32,
    #[serde(rename = "completed")]
    pub completed: bool,
    #[serde(rename = "canceled")]
    pub canceled: bool,
    #[serde(rename = "messages")]
    pub messages: Vec<JobMessage>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mailbox_cursor_date_to_db_string() {
        let cursor = MailboxCursor::Date {
            date_string: "2025-10-14".to_string(),
        };
        assert_eq!(cursor.to_db_string(), "date,2025-10-14");
    }

    #[test]
    fn test_mailbox_cursor_uid_to_db_string() {
        let cursor = MailboxCursor::Uid {
            last_seen_uid: 42,
            last_seen_datetime: 1234567890,
        };
        assert_eq!(cursor.to_db_string(), "uid,42,1234567890");
    }

    #[test]
    fn test_mailbox_cursor_date_from_db_string() {
        let s = "date,2025-10-14";
        assert_eq!(
            MailboxCursor::from_db_string(s).unwrap(),
            MailboxCursor::Date {
                date_string: "2025-10-14".to_string()
            }
        );
    }

    #[test]
    fn test_mailbox_cursor_uid_from_db_string() {
        let s = "uid,42,1234567890";
        assert_eq!(
            MailboxCursor::from_db_string(s).unwrap(),
            MailboxCursor::Uid {
                last_seen_uid: 42,
                last_seen_datetime: 1234567890
            }
        );
    }

    #[test]
    fn test_mailbox_cursor_invalid_from_db_string() {
        let s = "bad,format,here,extra";
        let result: Result<MailboxCursor, _> = MailboxCursor::from_db_string(s);
        assert!(result.is_err());
    }
}
